File: services/knowledge-hub/src/connectors/jira_connector.ts
/**
 * Jira connector — syncs epics, stories, and bugs from Atlassian Jira
 * using the REST API v3 (cloud) with Basic auth (email + API token).
 */

import {
  BaseConnector,
  ConnectorConfig,
  NormalizedEntity,
} from './base_connector';

// ────────────────────────────────────────────────────────────────────────────
// Types
// ────────────────────────────────────────────────────────────────────────────

/** Minimal representation of a Jira issue returned by the REST API. */
interface JiraIssue {
  id: string;
  key: string;
  self: string;
  fields: {
    summary: string;
    description?: string | null;
    status: { name: string };
    priority?: { name: string } | null;
    issuetype: { name: string };
    assignee?: { displayName: string; emailAddress?: string } | null;
    creator?: { displayName: string } | null;
    labels?: string[];
    fixVersions?: Array<{ name: string }>;
    sprint?: { name: string; state?: string } | null;
    updated: string;
    created: string;
    /* Epic link — custom field varies per instance; we also try parent. */
    parent?: { key: string } | null;
    [customField: string]: unknown;
  };
}

interface JiraSearchResponse {
  startAt: number;
  maxResults: number;
  total: number;
  issues: JiraIssue[];
}

// ────────────────────────────────────────────────────────────────────────────
// Connector
// ────────────────────────────────────────────────────────────────────────────

export class JiraConnector extends BaseConnector {
  private authHeader = '';

  constructor(config: ConnectorConfig) {
    super({
      ...config,
      sourceSystem: 'jira',
    });
  }

  // ── lifecycle ─────────────────────────────────────────────────────────

  async connect(): Promise<void> {
    const { email, apiToken } = this.config.auth;
    if (!email || !apiToken) {
      throw new Error('[JiraConnector] Missing auth.email or auth.apiToken');
    }

    this.authHeader =
      'Basic ' + Buffer.from(`${email}:${apiToken}`).toString('base64');

    // Validate credentials with a lightweight call
    await this.withRetry(async () => {
      const res = await fetch(this.buildUrl('/rest/api/3/myself'), {
        headers: { Authorization: this.authHeader, Accept: 'application/json' },
      });
      if (!res.ok) {
        throw new Error(`Jira auth check failed: ${res.status} ${res.statusText}`);
      }
    }, 'connect');

    this.connected = true;
  }

  async disconnect(): Promise<void> {
    this.authHeader = '';
    this.connected = false;
  }

  // ── data fetching ─────────────────────────────────────────────────────

  protected async *fetchRaw(
    watermark: string | null,
  ): AsyncGenerator<Record<string, unknown>> {
    const pageSize = 100;
    let startAt = 0;
    let total = Infinity;

    // JQL: fetch epics, stories, bugs; optionally filter by updated date
    let jql = 'issuetype in (Epic, Story, Bug) ORDER BY updated DESC';
    if (watermark) {
      // watermark is an ISO-8601 timestamp
      const jiraDate = watermark.replace('T', ' ').substring(0, 19);
      jql = `issuetype in (Epic, Story, Bug) AND updated >= "${jiraDate}" ORDER BY updated DESC`;
    }

    while (startAt < total) {
      const data = await this.withRetry(async () => {
        const url = this.buildUrl('/rest/api/3/search', {
          jql,
          startAt: String(startAt),
          maxResults: String(pageSize),
          fields:
            'summary,description,status,priority,issuetype,assignee,creator,' +
            'labels,fixVersions,sprint,updated,created,parent',
        });

        const res = await fetch(url, {
          headers: {
            Authorization: this.authHeader,
            Accept: 'application/json',
          },
        });
        if (!res.ok) {
          throw new Error(`Jira search failed: ${res.status}`);
        }
        return (await res.json()) as JiraSearchResponse;
      }, 'fetchRaw');

      total = data.total;

      for (const issue of data.issues) {
        yield issue as unknown as Record<string, unknown>;
      }

      startAt += data.issues.length;
      if (data.issues.length === 0) break;
    }
  }

  // ── normalisation ─────────────────────────────────────────────────────

  normalize(rawPayload: Record<string, unknown>): NormalizedEntity {
    const issue = rawPayload as unknown as JiraIssue;
    const f = issue.fields;
    const now = new Date().toISOString();

    return {
      source_system: 'jira',
      source_object_id: issue.key,
      entity_type: 'jira_issue',
      name: f.summary,
      description: this.buildDescription(f),
      owner: f.assignee?.displayName ?? f.creator?.displayName ?? null,
      status: this.mapStatus(f.status.name),
      source_url: `${this.config.baseUrl}/browse/${issue.key}`,
      last_synced_at: now,
      freshness_score: this.computeFreshness(now),
      eligible_skill_families_json: this.inferSkillFamilies(f),
    };
  }

  // ── private helpers ───────────────────────────────────────────────────

  /**
   * Build a human-readable description from Jira fields.
   */
  private buildDescription(f: JiraIssue['fields']): string {
    const parts: string[] = [];
    parts.push(`[${f.issuetype.name}] ${f.summary}`);
    if (f.priority?.name) parts.push(`Priority: ${f.priority.name}`);
    if (f.labels?.length) parts.push(`Labels: ${f.labels.join(', ')}`);
    if (f.sprint?.name) parts.push(`Sprint: ${f.sprint.name}`);
    if (f.fixVersions?.length) {
      parts.push(`Fix versions: ${f.fixVersions.map((v) => v.name).join(', ')}`);
    }
    if (f.parent?.key) parts.push(`Epic: ${f.parent.key}`);
    if (typeof f.description === 'string') {
      parts.push(`\n${f.description}`);
    }
    return parts.join(' | ');
  }

  /**
   * Map Jira status names to a normalised status string.
   */
  private mapStatus(jiraStatus: string): string {
    const s = jiraStatus.toLowerCase();
    if (['done', 'closed', 'resolved'].includes(s)) return 'done';
    if (['in progress', 'in review', 'in development'].includes(s)) return 'in_progress';
    if (['to do', 'open', 'backlog', 'new'].includes(s)) return 'todo';
    return s.replace(/\s+/g, '_');
  }

  /**
   * Heuristic: infer which skill families might consume this entity.
   */
  private inferSkillFamilies(f: JiraIssue['fields']): string[] {
    const families: string[] = [];
    const labels = (f.labels ?? []).map((l) => l.toLowerCase());

    if (labels.includes('support') || f.issuetype.name === 'Bug') {
      families.push('agent');
    }
    if (labels.includes('copy') || labels.includes('marketing')) {
      families.push('copywriting');
    }
    if (labels.includes('outreach')) {
      families.push('outreach');
    }
    // Product issues are generally useful across all families
    if (families.length === 0) {
      families.push('general');
    }
    return families;
  }
}

File: services/knowledge-hub/src/connectors/jira_connector.py
"""
Jira connector for DreamFi - syncs Jira issues to knowledge hub.

Implements:
- Email + API token authentication
- JQL-based filtering for issues
- Watermark-based incremental sync
- Freshness scoring (7-day half-life)
"""

import os
import requests
from typing import Dict, List, Any, AsyncGenerator, Optional
from datetime import datetime, timedelta, timezone
import hashlib
import json


class NormalizedEntity:
    """Normalized representation of a synced entity."""
    
    def __init__(self, **kwargs):
        self.source_system = kwargs.get('source_system', 'jira')
        self.source_object_id = kwargs.get('source_object_id')
        self.entity_type = kwargs.get('entity_type')
        self.name = kwargs.get('name')
        self.description = kwargs.get('description', '')
        self.owner = kwargs.get('owner')
        self.status = kwargs.get('status', 'active')
        self.source_url = kwargs.get('source_url')
        self.last_synced_at = kwargs.get('last_synced_at')
        self.freshness_score = kwargs.get('freshness_score', 1.0)
        self.eligible_skill_families_json = kwargs.get('eligible_skill_families_json', [])


class JiraConnector:
    """
    Connector for Jira API v3.
    
    Fetches issues, normalizes to canonical schema, tracks freshness.
    """
    
    def __init__(self, config: Optional[Dict[str, Any]] = None):
        self.config = config or {}
        self.email = os.getenv('JIRA_EMAIL', '')
        self.api_token = os.getenv('JIRA_API_TOKEN', '')
        self.base_url = os.getenv('JIRA_BASE_URL', 'https://jira.example.com')
        self.connected = False
        
    async def connect(self):
        """Establish connection (verify credentials)."""
        if not self.email or not self.api_token:
            raise ValueError("JIRA_EMAIL and JIRA_API_TOKEN required")
        self.connected = True
        
    async def disconnect(self):
        """Close connection."""
        self.connected = False
        
    async def fetchRaw(self, watermark: Optional[str] = None) -> AsyncGenerator[Dict[str, Any], None]:
        """
        Fetch issues via JQL, sorted by updated date descending.
        Yields raw Jira API payloads.
        """
        if not self.connected:
            await self.connect()
            
        # JQL: Get non-closed issues (Epic, Story, Bug types)
        jql = 'issuetype in (Epic, Story, Bug) AND status != Closed'
        
        # If watermark provided, fetch only updated-since
        if watermark:
            jql += f' AND updated >= "{watermark}"'
        
        jql += ' ORDER BY updated DESC'
        
        url = f'{self.base_url}/rest/api/3/search'
        auth = (self.email, self.api_token)
        start_at = 0
        page_size = 100
        
        while True:
            params = {
                'jql': jql,
                'startAt': start_at,
                'maxResults': page_size,
                'expand': 'changelog'
            }
            
            response = requests.get(url, params=params, auth=auth, timeout=10)
            if response.status_code != 200:
                raise Exception(f"Jira API error: {response.status_code}")
            
            data = response.json()
            
            for issue in data.get('issues', []):
                yield issue
            
            if start_at + page_size >= data.get('total', 0):
                break
                
            start_at += page_size
    
    def normalize(self, raw_payload: Dict[str, Any]) -> NormalizedEntity:
        """
        Normalize Jira issue → NormalizedEntity.
        Maps fields, computes freshness, infers skill families.
        """
        fields = raw_payload.get('fields', {})
        
        # Extract issue key and URL
        issue_key = raw_payload.get('key', '')
        source_url = f'{self.base_url}/browse/{issue_key}'
        
        # Extract owner (assignee)
        assignee = fields.get('assignee')
        owner = assignee.get('displayName') if assignee else 'unassigned'
        
        # Extract updated timestamp and compute freshness
        updated_str = fields.get('updated', datetime.utcnow().isoformat())
        updated = datetime.fromisoformat(updated_str.replace('Z', '+00:00'))
        age_days = (datetime.utcnow(timezone.utc) - updated).days if hasattr(updated, 'tzinfo') else (datetime.utcnow() - updated).days
        
        # Freshness: 2^(-age_days / 7) [7-day half-life]
        freshness_score = 2 ** (-max(0, age_days) / 7.0)
        
        # Infer skill families from labels and issue type
        labels = fields.get('labels', [])
        issue_type = fields.get('issuetype', {}).get('name', '').lower()
        
        skill_families = []
        label_keywords = ' '.join(labels).lower()
        
        if 'support' in label_keywords or 'bug' in issue_type:
            skill_families.append('agent')
        if 'copy' in label_keywords or 'marketing' in label_keywords:
            skill_families.append('copywriting')
        if 'kb' in label_keywords or 'knowledge' in label_keywords:
            skill_families.append('agent')
        
        # Default to agent if no matches
        if not skill_families:
            skill_families = ['agent']
        
        return NormalizedEntity(
            source_system='jira',
            source_object_id=issue_key,
            entity_type='jira_issue',
            name=fields.get('summary', ''),
            description=fields.get('description', {}).get('content', [{}])[0].get('text', '') if isinstance(fields.get('description'), dict) else '',
            owner=owner,
            status=fields.get('status', {}).get('name', 'unknown').lower(),
            source_url=source_url,
            last_synced_at=datetime.utcnow().isoformat(),
            freshness_score=freshness_score,
            eligible_skill_families_json=skill_families
        )
    
    async def sync(self, watermark: Optional[str] = None) -> Dict[str, Any]:
        """
        Sync issues: fetch → normalize → return.
        """
        await self.connect()
        entities = []
        
        async for raw in self.fetchRaw(watermark):
            entity = self.normalize(raw)
            entities.append(entity)
        
        await self.disconnect()
        
        return {
            'entities': entities,
            'result': {
                'entitiesSynced': len(entities),
                'connector': 'jira',
                'syncedAt': datetime.utcnow().isoformat()
            }
        }
